earth2studio/models/auto.py

```diff
@@ -14,85 +14,287 @@
+        by default {}
     """
 
-    def __init__(self, root: str, cache_dir: str | None = None):
+    def __init__(
+        self,
+        root: str,
+        fs: AbstractFileSystem | None = None,
+        cache: bool = True,
+        cache_options: dict = {},
+    ):
+
+        self.cache_options = cache_options.copy()
+        if "cache_storage" not in self.cache_options:
+            self.cache_options["cache_storage"] = self.default_cache()
+        if "expiry_time" not in self.cache_options:
+            self.cache_options["expiry_time"] = 31622400  # 1 year
+
         self.root = root
-        self.cache_dir = cache_dir
-        if not self.cache_dir:
-            # TODO: Update to earth2 studio after fix to
-            # https://github.com/NVIDIA/modulus/blob/main/modulus/utils/filesystem.py#L101
-            # Is upstreamed
-            default_cache = os.path.join(os.path.expanduser("~"), ".cache", "modulus")
-            self.cache_dir = os.environ.get("EARTH2STUDIO_CACHE", default_cache)
-        # Create cache dir if it doesn't exist
-        os.makedirs(self.cache_dir, exist_ok=True)
-
-    def get(self, path: str, recursive: bool = False, same_names: bool = False) -> str:
-        """Get a local path to the item at ``path``
+
+        if fs is not None:
+            self.fs = fs
+        elif root.startswith("ngc://models/"):
+            # Taken from Modulus file utils
+            # Strip ngc model url prefix
+            suffix = "ngc://models/"
+            # The regex check
+            pattern = re.compile(rf"{suffix}[\w-]+(/[\w-]+)?/[\w-]+@[A-Za-z0-9.]+")
+            if not pattern.match(root):
+                raise ValueError(
+                    "Invalid URL, should be of form ngc://models/<org_id/team_id/model_id>@<version>\n"
+                    + f" got {root}"
+                )
+            root = root.replace(suffix, "")
+            if len(root.split("@")[0].split("/")) == 3:
+                (org, team, model_version) = root.split("/", 2)
+                (model, version) = model_version.split("@", 1)
+            else:
+                (org, model_version) = root.split("/", 1)
+                (model, version) = model_version.split("@", 1)
+                team = None
+            if team:
+                self.root = f"https://api.ngc.nvidia.com/v2/models/{org}/{team}/{model}/versions/{version}/files/"
+            else:
+                self.root = f"https://api.ngc.nvidia.com/v2/models/{org}/{model}/versions/{version}/files/"
+            self.fs = HTTPFileSystem(block_size=2**10)
+        elif root.startswith("hf://"):
+            self.fs = HfFileSystem(target_options={"default_block_size": 2**20})
+        elif root.startswith("s3://"):
+            self.fs = s3fs.S3FileSystem(
+                anon=True,
+                client_kwargs={},
+                target_options={"default_block_size": 2**20},
+            )
```

Comment:
It has been my experience in the past that I have needed to pass `client_kwargs`. Not sure there is an elegant solution here, though.

Reply:
Yeah this assumes that its a public bucket (options based on settings used for s3 based data sources).

I didn't want to have file system specific stuff in the constructor of the package, so my solution was to permit users to just provide a initialized filestore. So in the case of a private S3 you would provide fs in the constructor.

This is borrowed from how Fsspec deals with file stores with caching systems.

Reply:
Good point, I missed that initially.